Canonical DateTime formats for java.time

azito-ct opened the pull request:
This PR ports `java.time` parsers and formatters from the [api-reference-scala](https://github.com/commercetools/commercetools-api-reference-scala/blob/fab9d1caf62293b6306fbf597df4725848442dca/src/main/scala/com/commercetools/api/CommercetoolsClient.scala#L89) repo.

These formats are designed to maintain high compatibility with Joda time and are currently used by [search-services](https://github.com/commercetools/search-services).

This PR does NOT change the formats used to encode/decode Json dates, but only point out places where the newer format could be considered to improve compatibility with Joda.
Further work should be carried on as part of the [java.time migration](https://commercetools.atlassian.net/browse/BEC-254) Backend Chapter ticket.

The motivation for this work is an issue encountered in PPS, where a date with a year containing more than 4 digits has been encoded with a format non compatible with ES (see [here](https://commercetools.slack.com/archives/CD0KL3DB3/p1746692639664929) for details).

satabin commented:
Thanks for this. Do you think it could be possible to have property roundtrip tests similar to what we have in https://github.com/commercetools/sphere-scala-libs/blob/main/json/json-core/src/test/scala/io/sphere/json/JodaJavaTimeCompat.scala ?

azito-ct commented:
@satabin Roundtrip property tests added alongside manual tests in https://github.com/commercetools/sphere-scala-libs/pull/668/commits/cd56c6f138f6b6fbf54b6dd857bcb39f8194c96e

satabin commented:
Nice, thanks. There has been some movement on the file though, and things were extracted, you'll need to rebase and solve the conflict.

azito-ct commented:
Rebased.

@yanns #670 conflicted with this PR. 
Since the intent behind the 2 PRs is similar but this one is a bit more wide in scope I reverted your changes.

Can you give it a look and raise any concern if you think this is not desirable?

yanns commented on util/src/main/scala/DateTimeFormats.scala:
```diff
@@ -0,0 +1,156 @@
+    *
+    * Parse a ZonedDateTime defaulting to UTC using the lenient date and time parsers
+    */
+  val lenientDateTimeParser: DateTimeFormatter =
```
Can we make all parsers private and only expose the functions?

yanns commented on json/json-core/src/main/scala/io/sphere/json/FromJSON.scala:
```diff
@@ -379,15 +380,48 @@ object FromJSON extends FromJSONInstances {
+  // similar to what the joda `DateTime` constructor accepts
+  // the accepted grammar for joda is described here: https://www.joda.org/joda-time/apidocs/org/joda/time/format/ISODateTimeFormat.html#dateTimeParser--
+  // this only supports the part where the date is specified
+  private val lenientInstantParser =
```
Why is this not in `DateTimeFormats`?

azito-ct replied:
Because my intent would be to use `DateTimeFormats. lenientDateTimeParser` instead as it should be equivalent (but a bit more flexible).

yanns commented on util/src/main/scala/DateTimeFormats.scala:
```diff
@@ -0,0 +1,156 @@
+    Try(
+      OffsetDateTime.parse(dateTime, lenientDateTimeParser)
+    )
+  def parseInstant(dateTime: String): Try[Instant] =
+    Try(
+      OffsetDateTime.parse(dateTime, lenientDateTimeParser).toInstant()
```
How is this different from https://github.com/commercetools/sphere-scala-libs/blob/3909c0d7e1fd4b70d8b0e2cd4c9db52cd9c4699b/json/json-core/src/main/scala/io/sphere/json/TimeParsers.scala#L11-L12 ?
Why are we not using the `lenientInstantParser`?
I guess that if I use the one, this will break the parsing I'm using in the GraphQL.

azito-ct replied:
The main difference between those parsers is that the one in `DateTimeFormats` supports years with more than 4 digit, is a bit more flexible in the parsing of time offsets.

If you replace the current instant parser with the one from `DateTimeFormats` as indicated [here](https://github.com/commercetools/sphere-scala-libs/blob/1ffc43a31b7860d853b848476192c42e585a9e4d/json/json-core/src/main/scala/io/sphere/json/FromJSON.scala#L412), all tests are green apart for the one that checks for failure for years way in the future.

In theory you should be able to swap to the parser in `DateTimeFormats` without any breakage as it is mostly a more flexible version of the current one.

yanns commented:
> This PR does NOT change the formats used to encode/decode Json dates, but only point out places where the newer format could be considered to improve compatibility with Joda.

If I understand correctly, you're adding new parsers but you don't use them right now.
What are the differences with the current ones? Why not directly changing the current ones?
I'm afraid that we'll never merge both approaches.

If you explain me the differences, I'm ready to tackle those with the current ones, avoiding the duplication in the first place.

yanns commented:
I'd really like to have one source of truth that I can also use in other places.

azito-ct commented:
> I'd really like to have one source of truth that I can also use in other places.

This is the same goal I have.
The parsers and formats I ported are the one used in search-services through [defined in api-reference-scala](https://github.com/commercetools/commercetools-api-reference-scala/blob/fab9d1caf62293b6306fbf597df4725848442dca/src/main/scala/com/commercetools/api/CommercetoolsClient.scala#L89).

The idea is to have a single authoritative parser/format source so that's we don't end up having issue due to difference in date formats.

> If I understand correctly, you're adding new parsers but you don't use them right now.
What are the differences with the current ones? Why not directly changing the current ones?

The difference is mostly in better handling of years with more than 4 digits which caused issues both in search-services as well as in PPS (some customers use dates in the year 10000 and some use negative years).

I would be in favour of directly changing the current parsers/formatters, but I cannot take up the task right now. 
My idea was to prepare the formats, pooint out the place where they can be used (see the comments in the PR's code), and leave the integration work to somebody in the Backend Chapter.

yanns commented:
OK then if it's ok, I can continue the work in this branch.

yanns reviewed (approved):
After having spent **time** on this, I don't think it's worth replicating the same limits with java time as we have with joda. So I'm fine merging this PR.
